common/persistence/sql/sqlShardPersistence.go

```diff
@@ -0,0 +1,322 @@
+// THE SOFTWARE.
+
+package sql
+
+import (
+	"fmt"
+	"time"
+
+	"database/sql"
+	"github.com/jmoiron/sqlx"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	sqlShardManager struct {
+		db *sqlx.DB
+	}
+
+	shardsRow struct {
+		ShardID                   int64     `db:"shard_id"`
+		Owner                     string    `db:"owner"`
+		RangeID                   int64     `db:"range_id"`
+		StolenSinceRenew          int64     `db:"stolen_since_renew"`
+		UpdatedAt                 time.Time `db:"updated_at"`
+		ReplicationAckLevel       int64     `db:"replication_ack_level"`
+		TransferAckLevel          int64     `db:"transfer_ack_level"`
+		TimerAckLevel             time.Time `db:"timer_ack_level"`
+		ClusterTransferAckLevel   []byte    `db:"cluster_transfer_ack_level"`
+		ClusterTimerAckLevel      []byte    `db:"cluster_timer_ack_level"`
+		DomainNotificationVersion int64     `db:"domain_notification_version"`
+	}
+)
+
+const (
+	createShardSQLQuery = `INSERT INTO shards 
+(shard_id, 
+owner, 
+range_id,
+stolen_since_renew,
+updated_at,
+replication_ack_level,
+transfer_ack_level,
+timer_ack_level,
+cluster_transfer_ack_level,
+cluster_timer_ack_level,
+domain_notification_version)
+VALUES
+(:shard_id, 
+:owner, 
+:range_id,
+:stolen_since_renew,
+:updated_at,
+:replication_ack_level,
+:transfer_ack_level,
+:timer_ack_level,
+:cluster_transfer_ack_level,
+:cluster_timer_ack_level,
+:domain_notification_version)`
+
+	getShardSQLQuery = `SELECT
+shard_id,
+owner,
+range_id,
+stolen_since_renew,
+updated_at,
+replication_ack_level,
+transfer_ack_level,
+timer_ack_level,
+cluster_transfer_ack_level,
+cluster_timer_ack_level,
+domain_notification_version
+FROM shards WHERE
+shard_id = ?
+`
+
+	updateShardSQLQuery = `UPDATE
+shards 
+SET
+shard_id = :shard_id,
+owner = :owner,
+range_id = :range_id,
+stolen_since_renew = :stolen_since_renew,
+updated_at = :updated_at,
+replication_ack_level = :replication_ack_level,
+transfer_ack_level = :transfer_ack_level,
+timer_ack_level = :timer_ack_level,
+cluster_transfer_ack_level = :cluster_transfer_ack_level,
+cluster_timer_ack_level = :cluster_timer_ack_level,
+domain_notification_version = :domain_notification_version
+WHERE
+shard_id = :shard_id AND
+range_id = :old_range_id
+`
+
+	lockShardSQLQuery   = `SELECT range_id FROM shards WHERE shard_id = ? FOR UPDATE`
+)
+
+func NewShardPersistence(username, password, host, port, dbName string) (persistence.ShardManager, error) {
+	var db, err = sqlx.Connect("mysql",
+		fmt.Sprintf(Dsn, username, password, host, port, dbName))
+	if err != nil {
+		return nil, err
+	}
+
+	return &sqlShardManager{
+		db: db,
+	}, nil
+}
+
+func (m *sqlShardManager) Close() {
+	if m.db != nil {
+		m.db.Close()
+	}
+}
+
+func (m *sqlShardManager) CreateShard(request *persistence.CreateShardRequest) error {
+	_, err := m.GetShard(&persistence.GetShardRequest{request.ShardInfo.ShardID})
```

Comment:
You can just blindly insert. If row exists you will receive errors as there will be conflict on primary key. Then you don't need the transaction and lock stuffs and it will be much simpler.